client/my-sites/domains/domain-management/list/bulk-edit-contact-info.jsx

```diff
@@ -0,0 +1,218 @@
+import { isFetchingUserSettings } from 'calypso/state/user-settings/selectors';
+
+import wp from 'calypso/lib/wp';
+
+const wpcom = wp.undocumented();
+
+/**
+ * Style dependencies
+ */
+import './list-all.scss';
+
+class BulkEditContactInfo extends React.Component {
+	static propTypes = {
+		handleSaveContactInfo: PropTypes.func.isRequired,
+		isDisabled: PropTypes.bool,
+		isSubmitting: PropTypes.bool,
+		onTransferLockOptOutChange: PropTypes.func.isRequired,
+		emailOnly: PropTypes.bool,
+	};
+
+	static defaultProps = {
+		isDisabled: false,
+		isSubmitting: false,
+		emailOnly: false,
+	};
+
+	state = {
+		hasSetContactDetailsFromCache: false,
+		contactDetails: {},
+		errorMessages: {},
+	};
+
+	componentDidMount() {
+		if ( isEmpty( this.props.userSettings ) ) {
+			this.props.fetchUserSettings();
+		}
+
+		if ( isEmpty( this.props.contactDetailsCache ) ) {
+			this.props.requestContactDetailsCache();
+		}
+
+		this.debouncedValidateContactDetails = debounce( this.validateContactDetails, 500 );
+	}
+
+	componentDidUpdate() {
+		const email = ! isEmpty( this.props.userSettings?.new_user_email )
+			? this.props.userSettings?.new_user_email
+			: this.props.userSettings?.user_email;
+
+		if ( ! isEmpty( email ) && isEmpty( this.state.contactDetails?.email ) ) {
+			this.updateDomainContactFields( { email: email } );
+		}
+
+		if (
+			! isEmpty( this.props.contactDetailsCache ) &&
+			false === this.state.hasSetContactDetailsFromCache
+		) {
+			this.setContactDetailsFromCache( this.props.contactDetailsCache );
+		}
+	}
+
+	isLoading() {
+		return (
+			isEmpty( this.props.contactDetailsCache ) ||
+			this.props.isRequestingContactDetailsCache ||
+			this.props.isFetchingUserSettings
+		);
+	}
+
+	setContactDetailsFromCache = ( data ) => {
+		delete data.email;
+		this.setState( { hasSetContactDetailsFromCache: true }, () => {
+			this.updateDomainContactFields( data );
+		} );
+	};
+
+	updateDomainContactFields = ( data ) => {
+		const newContactDetails = merge( {}, this.state.contactDetails, data );
+		this.setState( { contactDetails: newContactDetails } );
+		! isEmpty( newContactDetails ) && this.debouncedValidateContactDetails( newContactDetails );
+	};
+
+	onTransferLockOptOutChange = ( event ) => {
+		this.props.onTransferLockOptOutChange( event.target.checked );
+	};
+
+	validateContactDetails = ( contactDetails ) => {
+		if ( this.isLoading() ) {
+			return;
+		}
+
+		wpcom.validateDomainContactInformation( contactDetails, [], ( error, data ) => {
+			let errorMessages = ( data && data.messages ) || {};
+			if ( ! isEmpty( errorMessages ) ) {
+				errorMessages = mapValues( errorMessages, ( errors ) => {
+					return Array.isArray( errors ) ? errors.join( ' ' ) : errors;
+				} );
+			}
+
+			this.setState( {
+				errorMessages,
+			} );
+		} );
+	};
+
+	handleSaveContactInfo = () => this.props.handleSaveContactInfo( this.state.contactDetails );
+
+	renderTransferLockOptOut() {
```

Comment:
Minor: Since this render function is not particularly complicated, could it be moved to its own functional component? That would reduce the use of `this` in the function which would allow linters to notice any problems with its props before runtime.

Reply:
This is a really good point.  Thank you for the suggestion.